File: ios/Tradeguruelectrical/Services/TradeGuruAPI.swift
import Foundation

enum TradeGuruAPIError: Error {
    case noDeviceId
    case httpError(Int, String)
    case decodingError(String)
    case streamError(StreamErrorPayload)
    case networkError(Error)
}

enum TradeGuruAPI {

    // MARK: - Device Registration

    static func registerDevice() async throws -> String {
        var req = APIConfig.request("device/register", deviceId: "pending")
        let body: [String: String] = [
            "platform": APIConfig.platform,
            "app_version": APIConfig.appVersion,
            "locale": Locale.current.identifier,
            "timezone": TimeZone.current.identifier,
        ]
        req.httpBody = try JSONEncoder().encode(body)
        req.setValue("application/json", forHTTPHeaderField: "Content-Type")
        req.allHTTPHeaderFields?.removeValue(forKey: "X-Device-ID")

        let (data, response) = try await URLSession.shared.data(for: req)
        try validateResponse(response, data: data)

        let result = try JSONDecoder().decode(DeviceRegisterResponse.self, from: data)
        return result.deviceId
    }

    // MARK: - Chat (SSE Streaming)

    static func chat(
        messages: [[String: String]],
        mode: ThinkingMode,
        deviceId: String,
        jwt: String? = nil
    ) -> AsyncStream<StreamResult> {
        AsyncStream { continuation in
            Task {
                do {
                    var req = APIConfig.request("chat", deviceId: deviceId, jwt: jwt)
                    let body: [String: Any] = [
                        "messages": messages,
                        "mode": mode.rawValue,
                        "device_id": deviceId,
                        "platform": APIConfig.platform,
                    ]
                    req.httpBody = try JSONSerialization.data(withJSONObject: body)

                    let (bytes, response) = try await URLSession.shared.bytes(for: req)

                    guard let http = response as? HTTPURLResponse else {
                        continuation.yield(.error(StreamErrorPayload(code: "NO_RESPONSE", message: "No HTTP response", partial: nil)))
                        continuation.finish()
                        return
                    }

                    guard http.statusCode == 200 else {
                        let errorBody = await collectErrorBody(bytes)
                        continuation.yield(.error(StreamErrorPayload(
                            code: "HTTP_\(http.statusCode)",
                            message: errorBody,
                            partial: nil
                        )))
                        continuation.finish()
                        return
                    }

                    for await result in StreamParser.parse(lines: bytes.lines) {
                        continuation.yield(result)
                    }

                    continuation.finish()
                } catch {
                    continuation.yield(.error(StreamErrorPayload(
                        code: "NETWORK_ERROR",
                        message: error.localizedDescription,
                        partial: nil
                    )))
                    continuation.finish()
                }
            }
        }
    }

    // MARK: - Vision (Photo Analysis)

    static func chatVision(
        message: String,
        image: String,
        mode: ThinkingMode,
        deviceId: String,
        jwt: String? = nil
    ) -> AsyncStream<StreamResult> {
        AsyncStream { continuation in
            Task {
                do {
                    var req = APIConfig.request("chat/vision", deviceId: deviceId, jwt: jwt)
                    let body: [String: Any] = [
                        "message": message,
                        "image": image,
                        "mode": mode.rawValue,
                        "device_id": deviceId,
                        "platform": APIConfig.platform,
                    ]
                    req.httpBody = try JSONSerialization.data(withJSONObject: body)

                    let (bytes, response) = try await URLSession.shared.bytes(for: req)

                    guard let http = response as? HTTPURLResponse, http.statusCode == 200 else {
                        continuation.yield(.error(StreamErrorPayload(code: "HTTP_ERROR", message: "Vision request failed", partial: nil)))
                        continuation.finish()
                        return
                    }

                    for await result in StreamParser.parse(lines: bytes.lines) {
                        continuation.yield(result)
                    }

                    continuation.finish()
                } catch {
                    continuation.yield(.error(StreamErrorPayload(code: "NETWORK_ERROR", message: error.localizedDescription, partial: nil)))
                    continuation.finish()
                }
            }
        }
    }

    // MARK: - Audio

    static func transcribe(audioData: Data, mimeType: String, deviceId: String, jwt: String? = nil) async throws -> String {
        let boundary = "TradeGuru-\(UUID().uuidString)"
        var req = APIConfig.request("audio/transcribe", deviceId: deviceId, jwt: jwt)
        req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()
        let ext = mimeType.contains("wav") ? "wav" : "m4a"
        body.append("--\(boundary)\r\nContent-Disposition: form-data; name=\"audio\"; filename=\"recording.\(ext)\"\r\nContent-Type: \(mimeType)\r\n\r\n".data(using: .utf8) ?? Data())
        body.append(audioData)
        body.append("\r\n--\(boundary)--\r\n".data(using: .utf8) ?? Data())
        req.httpBody = body

        let (data, response) = try await URLSession.shared.data(for: req)
        try validateResponse(response, data: data)

        let result = try JSONDecoder().decode(TranscribeResponse.self, from: data)
        return result.text
    }

    static func speak(text: String, voice: String = "nova", deviceId: String, jwt: String? = nil) async throws -> Data {
        var req = APIConfig.request("audio/speech", deviceId: deviceId, jwt: jwt)
        let body = ["text": text, "voice": voice]
        req.httpBody = try JSONEncoder().encode(body)

        let (data, response) = try await URLSession.shared.data(for: req)
        try validateResponse(response, data: data)
        return data
    }

    // MARK: - File Upload

    static func uploadFile(fileData: Data, fileName: String, mimeType: String, deviceId: String, jwt: String? = nil) async throws -> FileUploadResponse {
        let boundary = "TradeGuru-\(UUID().uuidString)"
        var req = APIConfig.request("files/upload", deviceId: deviceId, jwt: jwt)
        req.setValue("multipart/form-data; boundary=\(boundary)", forHTTPHeaderField: "Content-Type")

        var body = Data()
        body.append("--\(boundary)\r\nContent-Disposition: form-data; name=\"file\"; filename=\"\(fileName)\"\r\nContent-Type: \(mimeType)\r\n\r\n".data(using: .utf8) ?? Data())
        body.append(fileData)
        body.append("\r\n--\(boundary)--\r\n".data(using: .utf8) ?? Data())
        req.httpBody = body

        let (data, response) = try await URLSession.shared.data(for: req)
        try validateResponse(response, data: data)
        return try JSONDecoder().decode(FileUploadResponse.self, from: data)
    }

    // MARK: - Rating & Feedback

    static func rate(responseId: String, stars: Int, mode: ThinkingMode, comment: String? = nil, deviceId: String, jwt: String? = nil) async throws {
        var req = APIConfig.request("rating", deviceId: deviceId, jwt: jwt)
        var body: [String: Any] = [
            "response_id": responseId,
            "stars": stars,
            "mode": mode.rawValue,
        ]
        if let comment { body["comment"] = comment }
        req.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: req)
        try validateResponse(response, data: data)
    }

    static func feedback(responseId: String, reason: String, detail: String? = nil, mode: ThinkingMode, deviceId: String, jwt: String? = nil) async throws {
        var req = APIConfig.request("feedback", deviceId: deviceId, jwt: jwt)
        var body: [String: Any] = [
            "response_id": responseId,
            "reason": reason,
            "mode": mode.rawValue,
        ]
        if let detail { body["detail"] = detail }
        req.httpBody = try JSONSerialization.data(withJSONObject: body)

        let (data, response) = try await URLSession.shared.data(for: req)
        try validateResponse(response, data: data)
    }

    // MARK: - Helpers

    private static func validateResponse(_ response: URLResponse, data: Data) throws {
        guard let http = response as? HTTPURLResponse else {
            throw TradeGuruAPIError.httpError(0, "No HTTP response")
        }
        guard (200...299).contains(http.statusCode) else {
            let body = String(data: data, encoding: .utf8) ?? "Unknown error"
            throw TradeGuruAPIError.httpError(http.statusCode, body)
        }
    }

    private static func collectErrorBody(_ bytes: URLSession.AsyncBytes) async -> String {
        var collected = ""
        for await line in bytes.lines {
            collected += line
            if collected.count > 500 { break }
        }
        return collected.isEmpty ? "Unknown error" : collected
    }
}

// MARK: - Response Types

nonisolated struct DeviceRegisterResponse: Codable {
    let deviceId: String
    enum CodingKeys: String, CodingKey {
        case deviceId = "device_id"
    }
}

nonisolated struct TranscribeResponse: Codable {
    let text: String
}

nonisolated struct FileUploadResponse: Codable {
    let id: String
    let filename: String
}

File: android/app/src/main/java/com/tradeguru/electrical/services/TradeGuruAPI.kt
package com.tradeguru.electrical.services

import com.google.gson.Gson
import com.tradeguru.electrical.models.ThinkingMode
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.flow
import kotlinx.coroutines.flow.flowOn
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.MultipartBody
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import java.io.IOException
import java.util.Locale
import java.util.TimeZone
import java.util.concurrent.TimeUnit

object TradeGuruAPI {
    private val client = OkHttpClient.Builder()
        .readTimeout(5, TimeUnit.MINUTES)
        .connectTimeout(30, TimeUnit.SECONDS)
        .build()
    private val gson = Gson()

    suspend fun registerDevice(): String = withContext(Dispatchers.IO) {
        val localeId = Locale.getDefault().toLanguageTag().ifEmpty { "en" }
        val tz = TimeZone.getDefault()
        val hours = tz.rawOffset / 3600000
        val sign = if (hours >= 0) "+" else ""
        val timezoneId = "GMT$sign$hours"

        val body = mapOf(
            "platform" to APIConfig.PLATFORM,
            "app_version" to APIConfig.APP_VERSION,
            "locale" to localeId,
            "timezone" to timezoneId
        )
        val jsonBody = gson.toJson(body)
            .toRequestBody("application/json".toMediaType())

        val request = Request.Builder()
            .url(APIConfig.url("device/register"))
            .post(jsonBody)
            .addHeader("Content-Type", "application/json")
            .build()

        val response = client.newCall(request).execute()
        val responseBody = response.body?.string()
            ?: throw IOException("Empty response body")
        if (!response.isSuccessful) {
            throw IOException("HTTP ${response.code}: $responseBody")
        }
        gson.fromJson(responseBody, DeviceRegisterResponse::class.java).deviceId
    }

    fun chat(
        messages: List<ApiMessage>,
        mode: ThinkingMode,
        deviceId: String,
        jwt: String? = null
    ): Flow<StreamResult> = flow {
        val body = mapOf(
            "messages" to messages,
            "mode" to mode.value,
            "device_id" to deviceId,
            "platform" to APIConfig.PLATFORM
        )
        val jsonBody = gson.toJson(body)
            .toRequestBody("application/json".toMediaType())

        val request = APIConfig.requestBuilder("chat", deviceId = deviceId, jwt = jwt)
            .post(jsonBody)
            .build()

        val call = client.newCall(request)
        var response: okhttp3.Response? = null
        try {
            response = withContext(Dispatchers.IO) { call.execute() }
            if (!response.isSuccessful) {
                val errorBody = response.body?.string() ?: "Unknown error"
                emit(StreamResult.Error(StreamErrorPayload(
                    code = "HTTP_${response.code}",
                    message = errorBody
                )))
                return@flow
            }

            val source = response.body?.source()
                ?: throw IOException("Empty body")
            var currentEvent = ""
            var currentData = ""
            while (!source.exhausted()) {
                ensureActive()
                val line = withContext(Dispatchers.IO) { source.readUtf8Line() } ?: break
                if (line.startsWith("event: ")) {
                    currentEvent = line.removePrefix("event: ").trim()
                } else if (line.startsWith("data: ")) {
                    currentData = line.removePrefix("data: ").trim()
                } else if (line.isEmpty() && currentEvent.isNotEmpty()) {
                    emit(StreamParser.parse(currentEvent, currentData))
                    currentEvent = ""
                    currentData = ""
                }
            }
        } finally {
            call.cancel()
        }
    }.flowOn(Dispatchers.IO)

    fun chatVision(
        message: String,
        image: String,
        mode: ThinkingMode,
        deviceId: String,
        jwt: String? = null
    ): Flow<StreamResult> = flow {
        val body = mapOf(
            "message" to message,
            "image" to image,
            "mode" to mode.value,
            "device_id" to deviceId,
            "platform" to APIConfig.PLATFORM
        )
        val jsonBody = gson.toJson(body)
            .toRequestBody("application/json".toMediaType())

        val request = APIConfig.requestBuilder("chat/vision", deviceId = deviceId, jwt = jwt)
            .post(jsonBody)
            .build()

        val call = client.newCall(request)
        var response: okhttp3.Response? = null
        try {
            response = withContext(Dispatchers.IO) { call.execute() }
            if (!response.isSuccessful) {
                val errorBody = response.body?.string() ?: "Unknown error"
                emit(StreamResult.Error(StreamErrorPayload(
                    code = "HTTP_${response.code}",
                    message = errorBody
                )))
                return@flow
            }

            val source = response.body?.source()
                ?: throw IOException("Empty body")
            var currentEvent = ""
            var currentData = ""
            while (!source.exhausted()) {
                ensureActive()
                val line = withContext(Dispatchers.IO) { source.readUtf8Line() } ?: break
                if (line.startsWith("event: ")) {
                    currentEvent = line.removePrefix("event: ").trim()
                } else if (line.startsWith("data: ")) {
                    currentData = line.removePrefix("data: ").trim()
                } else if (line.isEmpty() && currentEvent.isNotEmpty()) {
                    emit(StreamParser.parse(currentEvent, currentData))
                    currentEvent = ""
                    currentData = ""
                }
            }
        } finally {
            call.cancel()
        }
    }.flowOn(Dispatchers.IO)

    suspend fun uploadFile(
        fileData: ByteArray,
        fileName: String,
        mimeType: String,
        deviceId: String,
        jwt: String? = null
    ): FileUploadResponse = withContext(Dispatchers.IO) {
        val requestBody = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart(
                "file", fileName,
                fileData.toRequestBody(mimeType.toMediaType())
            )
            .build()

        val request = APIConfig.requestBuilder("files/upload", deviceId = deviceId, jwt = jwt)
            .post(requestBody)
            .removeHeader("Content-Type")
            .build()

        val response = client.newCall(request).execute()
        val body = response.body?.string() ?: throw IOException("Empty response body")
        if (!response.isSuccessful) {
            throw IOException("HTTP ${response.code}: $body")
        }
        gson.fromJson(body, FileUploadResponse::class.java)
    }

    suspend fun transcribe(
        audioData: ByteArray,
        mimeType: String,
        deviceId: String,
        jwt: String? = null
    ): String = withContext(Dispatchers.IO) {
        val ext = if (mimeType.contains("wav")) "wav" else "m4a"
        val requestBody = MultipartBody.Builder()
            .setType(MultipartBody.FORM)
            .addFormDataPart(
                "audio", "recording.$ext",
                audioData.toRequestBody(mimeType.toMediaType())
            )
            .build()

        val request = APIConfig.requestBuilder("audio/transcribe", deviceId = deviceId, jwt = jwt)
            .post(requestBody)
            .removeHeader("Content-Type")
            .build()

        val response = client.newCall(request).execute()
        val body = response.body?.string() ?: throw IOException("Empty response body")
        if (!response.isSuccessful) {
            throw IOException("HTTP ${response.code}: $body")
        }
        gson.fromJson(body, TranscribeResponse::class.java).text
    }

    suspend fun speak(
        text: String,
        voice: String = "nova",
        deviceId: String,
        jwt: String? = null
    ): ByteArray = withContext(Dispatchers.IO) {
        val body = mapOf("text" to text, "voice" to voice)
        val jsonBody = gson.toJson(body)
            .toRequestBody("application/json".toMediaType())

        val request = APIConfig.requestBuilder("audio/speech", deviceId = deviceId, jwt = jwt)
            .post(jsonBody)
            .build()

        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            val errorBody = response.body?.string() ?: "Unknown error"
            throw IOException("HTTP ${response.code}: $errorBody")
        }
        response.body?.bytes() ?: throw IOException("Empty response body")
    }

    suspend fun rate(
        responseId: String,
        stars: Int,
        mode: ThinkingMode,
        deviceId: String,
        comment: String? = null,
        jwt: String? = null
    ): Unit = withContext(Dispatchers.IO) {
        val bodyMap = mutableMapOf<String, Any>(
            "response_id" to responseId,
            "stars" to stars,
            "mode" to mode.value
        )
        if (comment != null) bodyMap["comment"] = comment
        val jsonBody = gson.toJson(bodyMap)
            .toRequestBody("application/json".toMediaType())

        val request = APIConfig.requestBuilder("rating", deviceId = deviceId, jwt = jwt)
            .post(jsonBody)
            .build()

        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            val errorBody = response.body?.string() ?: "Unknown error"
            throw IOException("HTTP ${response.code}: $errorBody")
        }
    }

    suspend fun feedback(
        responseId: String,
        reason: String,
        mode: ThinkingMode,
        deviceId: String,
        detail: String? = null,
        jwt: String? = null
    ): Unit = withContext(Dispatchers.IO) {
        val bodyMap = mutableMapOf<String, Any>(
            "response_id" to responseId,
            "reason" to reason,
            "mode" to mode.value
        )
        if (detail != null) bodyMap["detail"] = detail
        val jsonBody = gson.toJson(bodyMap)
            .toRequestBody("application/json".toMediaType())

        val request = APIConfig.requestBuilder("feedback", deviceId = deviceId, jwt = jwt)
            .post(jsonBody)
            .build()

        val response = client.newCall(request).execute()
        if (!response.isSuccessful) {
            val errorBody = response.body?.string() ?: "Unknown error"
            throw IOException("HTTP ${response.code}: $errorBody")
        }
    }
}
